qiskit/aqua/algorithms/amplitude_amplifiers/grover.py

```diff
@@ -111,192 +128,274 @@ def __init__(self,
 
-    @property
-    def qc_amplitude_amplification_iteration(self):
-        """ qc amplitude amplification iteration """
-        if self._qc_aa_iteration is None:
-            self._qc_aa_iteration = QuantumCircuit()
-            self._qc_aa_iteration += self._oracle.circuit
-            self._qc_aa_iteration += self._diffusion_circuit
-        return self._qc_aa_iteration
-
-    def _run_with_existing_iterations(self):
+    def _run_experiment(self, power):
+        """Run a grover experiment for a given power of the Grover operator."""
         if self._quantum_instance.is_statevector:
-            qc = self.construct_circuit(measurement=False)
+            qc = self.construct_circuit(power, measurement=False)
             result = self._quantum_instance.execute(qc)
-            complete_state_vec = result.get_statevector(qc)
-            variable_register_density_matrix = get_subsystem_density_matrix(
-                complete_state_vec,
-                range(len(self._oracle.variable_register), qc.width())
-            )
-            variable_register_density_matrix_diag = np.diag(variable_register_density_matrix)
-            max_amplitude = max(
-                variable_register_density_matrix_diag.min(),
-                variable_register_density_matrix_diag.max(),
-                key=abs
-            )
-            max_amplitude_idx = \
-                np.where(variable_register_density_matrix_diag == max_amplitude)[0][0]
-            top_measurement = np.binary_repr(max_amplitude_idx, len(self._oracle.variable_register))
+            statevector = result.get_statevector(qc)
+            num_bits = len(self._grover_operator.reflection_qubits)
+            # trace out work qubits
+            if qc.width() != num_bits:
+                rho = get_subsystem_density_matrix(
+                    statevector,
+                    range(num_bits, qc.width())
+                )
+                statevector = np.diag(rho)
+            max_amplitude = max(statevector.max(), statevector.min(), key=abs)
+            max_amplitude_idx = np.where(statevector == max_amplitude)[0][0]
+            top_measurement = np.binary_repr(max_amplitude_idx, num_bits)
+
         else:
-            qc = self.construct_circuit(measurement=True)
+            qc = self.construct_circuit(power, measurement=True)
             measurement = self._quantum_instance.execute(qc).get_counts(qc)
             self._ret['measurement'] = measurement
             top_measurement = max(measurement.items(), key=operator.itemgetter(1))[0]
 
         self._ret['top_measurement'] = top_measurement
-        oracle_evaluation, assignment = self._oracle.evaluate_classically(top_measurement)
-        return assignment, oracle_evaluation
 
-    def construct_circuit(self, measurement=False):
-        """
-        Construct the quantum circuit
+        return self.post_processing(top_measurement), self.is_good_state(top_measurement)
+
+    def is_good_state(self, bitstr: str) -> bool:
```

Comment:
This is intended to be a public method or should it be annotated to be private?

Reply:
The `is_good_state` argument in the general case should be a callable used to determine whether a measurement is a good state or not. For convenience, we support also passing a list of int and statevector, but it is still used as callable. It could be private, but if we leave it public users can check what their callable is, given the input and verify the result etc.

Reply:
The use with the callable seems odd - I guess in this case the user of the code supplying the callable and the user who wants to determine if its a good state are different users. If its the same user they already know how to do this since its really the callback, as far as I could tell, that was the code that actually did this in this case. I can see it could be public, the callable bit seemed to add a weird twist for me, hence the comment.

Reply:
Hm I see, yeah it doesn't really need to be public, it could also be private. The idea was to let users check different states they thought might be good or bad, without having to implement the logic that we're implementing anyways. E.g.
```
grover = Grover(..., is_good_state=[0, 2, 10], ...)
grover.is_good_state('0110100110')
```
but that's not a required feature (and might not be _that_ useful)